readme.md
# libxm7

A C library to play XM (and MOD) modules on Nintendo DS using only ARM7
resources (thus freeing your ARM9 from that task).

Homepage: http://wcms.teleion.it/users/cgq/nds/libxm7/

This is released as-is, under the MIT license.

- Demos of [version 0.81](https://www.youtube.com/watch?v=DOTtW3GKVTs) and
  [version 0.93](https://vimeo.com/3364671). Thanks, Raina!

- [AngunaDS](http://www.tolberts.net/anguna/) (the *first* homebrew featuring
  libXM7)

[Changelog](./changelog.txt)

## libXM7 Replay features

The current version of libXM7 (1.06) can play XMs featuring the following
characteristics:

- Up to 16 channels.
- Distinct volume and effect column for each channel.
- Module length up to 256 patterns.
- Module can contain up to 256 different patterns of variable length from 1 to
  256 lines.
- Up to 128 instruments, each one of which:

  - Has up to 16 different 8 or 16 bit samples, each one having a forward
    (normal) or a ping-pong loop optionally.
  - May have a volume envelope with up to 12 points, optionally with a sustain
    point and/or a loop, plus a release (fade out) rate.
  - May have a panning envelope with up to 12 points, optionally with a sustain
    point and/or a loop.
  - May have a vibrato definition with speed, depth and sweep.

- Both *linear frequency table* mode (XM native) and *Amiga frequency table*
  mode (MOD native) are supported for completeness, even if the latter is very
  uncommon nowadays

libXM7 has **complete** effect support. For reference, these are the effects:

- `0xy` Arpeggio
- `1xx` Portamento up (M)
- `2xx` Portamento down (M)
- `3xx` Portamento to note (M)
- `4xy` Vibrato (M)
- `5xy` Portamento to note with volume slide (M)
- `6xy` Vibrato with volume slide (M)
- `7xy` Tremolo (M)
- `8xx` Set note panning
- `9xx` Set sample offset (M)
- `Axy` Slide note volume up or down (M)
- `Bxx` Jump to position
- `Cxx` Set note volume
- `Dxx` Pattern break
- `E1x` Fine portamento up (M)
- `E2x` Fine portamento down (M)
- `E3x` Glissando control
- `E4x` Vibrato control
- `E5x` Instrument finetune
- `E6x` Pattern loop
- `E7x` Tremolo control
- `E8x` Set note panning
- `E9x` Retrig note
- `EAx` Slide note fine volume up (M)
- `EBx` Slide note fine volume down (M)
- `ECx` Note cut
- `EDx` Note delay
- `EEx` Pattern delay
- `Fxx` Set song speed
- `Gxx` Set global volume
- `Hxy` Slide global volume up or down (M)
- `Kxx` Key off
- `Lxx` Set envelope position
- `Pxy` Panning slide (M)
- `Rxy` Retrig note with volume slide (M)
- `Txy` Tremor (M)
- `X1x` Extra fine portamento up (M)
- `X2x` Extra fine portamento down (M)

(M) Means that this effect has "memory".

The volume column effects support is also **complete**. For your reference the
effects follows:

- `xx` Set note volume (`0x10-0x50`)
- `Dx` Slide note volume down (`0x60-0x6F`)
- `Ux` Slide note volume up (`0x70-0x7F`)
- `dx` Slide note fine volume down (`0x80-0x8F`)
- `ux` Slide note fine volume up (`0x90-0x9F`)
- `Sx` Set vibrato speed (`0xA0-0xAF`)
- `Vx` Vibrato (`0xB0-0xBF`)
- `Px` Set note panning (`0xC0-0xCF`)
- `Lx` Slide note panning left (`0xD0-0xDF`)
- `Rx` Slide note panning right (`0xE0-0xEF`)
- `Mx` Portamento to note (`0xF0-0xFF`)

There's a very good document explaining XM effects (and volume column effects)
usage [here](https://milkytracker.org/docs/MilkyTracker.html), on MilkyTracker
website.

The library can play also MODs featuring the following characteristics:

- `M.K.`, `M!K!` and `FLT4` module subtypes 4 channels modules
- `OCTA`, `CD81` and `FLT8` (\*) module subtypes 8 channels modules
- `TDZx` module subtype 1 to 3 channels modules
- `xCHN` module subtype 2 to 9 channels modules
- `xxCH` module subtype 10 to 16 channels modules
- Distinct effect column for each channel
- Module length up to 128 patterns
- Module can contain up to 256 different patterns of fixed length of 64 lines
  each.
- Up to 31 instruments, one 8 bits sample each, having a forward (normal) loop
  optionally.

Of course libXM7 has **complete** MOD effect support (because they're a subset of
XM effects).

(\*) Yeah, I mean `FLT8` really.

## Getting the most of your tune

Since the library doesn't manipulate the samples data directly but simply uses
the Nintendo DS hardware 16 digital channels to replay samples, there are some
things you should be aware of when composing tunes especially for this system.
First of all, remember that hardware mixing among the channels on a Nintendo DS
will result in a 10bits/32kHz stereo sound being reproduced. It's not poor
quality, but it isn't exactly CD quality either. In the mixing process the
samples don't get resampled when pitched so make sure you disable this feature
in your tracker (*sample interpolation*) if you want to simulate the output you
would get on the DS as better as possible.

The DS hardware also wants all the figures to be in 32bit boundaries, which
means that the samples should conform to some rules:

- Length of samples *without loop* should be *multiple* of 4 for 8 bits samples
  and should be even for 16 bits samples.

- The length of the **non-repeating** part of samples *with loops* (of any kind)
  and the length of the **repeating** part of samples *with forward loops*
  should be as stated above.

- Length of the **repeating** part of 8 bits samples *with ping-pong loops*
  should be *even*. No problems with *ping-pong loops* of 16 bits samples (this
  happens because DS hardware doesn't support ping-pong loops, so these loops
  are converted into forward loops by the loading routines).

A side effect (*detuning*) is possible when these rules are ignored, and it
could be perceivable on samples with loops, especially when the repeating part
of the sample is very short.

Finally, it's impossible for the DS hardware to start playing a sample that has
a loop from inside the loop from other position but the beginning. This means
that you should put some care in preparing your samples having loops if you plan
to use the `9xx` (*Set sample offset*) effect on them. Anyway there are no
problems if you simply plan to skip samples of the **non-repeating** part of a
sample with loop.

## How to use libXM7 files

The library consists of one header file (`libxm7.h`) and two archive files
(`libxm77.a` and `libxm79.a`). You should add the path with the header file
to your include paths, say the `include/libxm7` directory. On the other hand,
you should add the `libs` directory to your linker paths. Then you should
modify the makefiles for your project: locate the line where the libraries are
listed (`LIBS := -l<library> -l<library>`) and add `-lxm77` in the ARM7
makefile and `LIBDIRS := path/to/libxm7/libs`. Add also `-lmx79` in the ARM9
makefile if you plan to use both libXM7 library parts. Make sure that `-lxm77`
appears *before* `-lnds7` and that `-lxm79` appears *before* `-lnds9`.

## How to use libXM7 ARM7 functions

First of all, you should include `libxm7.h` in your ARM7 source file. Doing it
will give you access to the defined functions:

```c
void XM7_Initialize(void)
```

This function should be called just once, before calling any other function. It
performs some important necessary initialization.

```c
void XM7_PlayModule(XM7_ModuleManager_Type* TheModule)
```

This function makes the module start playing. It doesn't turn on the speakers,
nor does it set the master volume: all that is up to you. The following
parameter should be passed to the function: a pointer to a
`XM7_ModuleManager_Type` structure where an XM module should have already been
loaded.

```c
void XM7_PlayModuleFromPos(XM7_ModuleManager_Type* TheModule, u8 position)
```

This function makes the module start playing from a given position. It works
like the previous one but requires one more parameter: the position (in the
*Pattern Order Table*) where the replay should start. Using this function and
tracking XMs in 'sections' you can actually pack many tunes sharing the same
instruments in a single XM file. When it's required to switch to a different
tune you would then simply stop the module and play the same module again just
starting from a different position.

```c
void XM7_StopModule(void)
```

This function stops the module and it abruptly interrupts every sample of the
module being played.

## How to use libXM7 ARM9 functions

First, include `libxm7.h` in your ARM9 source file. The defined functions are:

```c
u16 XM7_LoadXM(XM7_ModuleManager_Type*, const void*)
```

This function loads an XM into the `XM7_ModuleManager_Type` structure that allows
the player to work. Both parameters are pointers; the first one should point to
an already allocated structure where this function will load the XM module,
whereas the second is the pointer to a copy in memory of a whole XM file. This
function uses `malloc()` to allocate space for patterns, instruments and
samples into the heap. Unlike the other functions, this function does return a
value, which is 0 when the loading is successful and a different value when the
loading has a different outcome. Error codes are discussed further on.

```c
u16 XM7_LoadMOD(XM7_ModuleManager_Type*, const void*)
```

This function loads a MOD into the `XM7_ModuleManager_Type` structure that
allows the player to work. Both parameters are pointers; the first one should
point to an already allocated structure where this function will load the
module, whereas the second is the pointer to a copy in memory of a whole MOD
file. This function uses `malloc()` to allocate space for patterns,
instruments and samples into the heap. Unlike the other functions, this function
does return a value, which is 0 when the loading is successful and a different
value when the loading has a different outcome. Error codes are discussed
further on.

```c
void XM7_UnloadXM(XM7_ModuleManager_Type*)
```

This function frees all the allocated memory thus unloading the module. However,
it doesn't deallocate the `XM7_ModuleManager_Type` structure.

```c
void XM7_UnloadMOD(XM7_ModuleManager_Type*)
```

This function is simply an alias of the `XM7_UnloadXM()` function, for your
convenience.

```c
void XM7_SetReplayStyle(XM7_ModuleManager_Type* Module, XM7_ReplayStyles style)
```

This function sets some parameters that affect the way the module will be
reproduced, mainly because there are some differencies in some effect behaviour
if used in XM or MOD modules. Actually this simply affects the effect `0xy`
(*Arpeggio*) and activates/deactivates the 'on-the-fly sample change' feature.
For your convenience some constants are defined:

```c
typedef enum {
    // Replay style flags:
    XM7_REPLAY_STYLE_XM_PLAYER           = 0x00,
    XM7_REPLAY_STYLE_MOD_PLAYER          = 0x01,
    XM7_REPLAY_ONTHEFLYSAMPLECHANGE_FLAG = 0x02,

    // Currently an alias of XM7_REPLAY_STYLE_XM_PLAYER, the default for XM
    // modules.
    XM7_REPLAY_STYLE_FT2 = 0x00,

    // The default for MOD modules
    XM7_REPLAY_STYLE_PT  = (XM7_REPLAY_STYLE_MOD_PLAYER | XM7_REPLAY_ONTHEFLYSAMPLECHANGE_FLAG)
} XM7_ReplayStyles;
```

```c
void XM7_SetPanningStyle(XM7_ModuleManager_Type* Module, XM7_PanningStyles style,
                         XM7_PanningDisplacementStyles displacement)
```

This function configures how the panning will be managed in the reproduction. XM
format has panning effects, instrument panning settings and even panning
envelopes, whereas MOD format has virtually none and originally (on Amiga
machines) mapped channels directly to the left speaker OR to the right speaker
only, following the rule that the first channel goes right and the second goes
left, then the opposite for the next two and swap again for the fifth and sixth
and so on (R-L-L-R-R-L-L-R ...). For your convenience again there are some
defined constants for the style:

```c
typedef enum {
    // Panning will be driven by the module. Default for XMs.
    XM7_PANNING_TYPE_NORMAL = 0x00,
    // Panning will be driven by the channel number. Default for MODs.
    XM7_PANNING_TYPE_AMIGA  = 0x01
} XM7_PanningStyles;
```

Then, if you select `XM7_PANNING_TYPE_AMIGA`, you can specity a value for the
displacement, this will be the 'distance' from the originally assigned speaker,
maximum value is 127 (which actually will completely swap channels...). The
following common constants are defined:

```c
// Panning displacement types:
typedef enum {
    // The panning as it was meant on Amiga.
    XM7_HARD_PANNING_DISPLACEMENT    = 0,
    // All the channels will be centered.
    XM7_MONO_PANNING_DISPLACEMENT    = 64,

    // Default. Panning mix: 1/3 of the left channels volume goes to the right
    // speaker and 1/3 of the right channels volume goes to the left speaker.
    // Quite common setting among MOD players.
    XM7_DEFAULT_PANNING_DISPLACEMENT = 42 // 42 = ~127/3 = panning 1/3 + 2/3
} XM7_PanningDisplacementStyles;
```

## `XM7_LoadXM()` and `XM7_LoadMOD()` error codes

Here's the list of the error code defines for your reference. They are all fatal
errors; the loading will be interrupted and it won't be possible to play the
module. The `XM7_UnloadXM()` function (or its alias `XM7_UnloadMOD()`)
should be called anyway to free the already allocated memory when the error code
is greater than `0x07`.

```c
typedef enum {
    XM7_NO_ERROR                           = 0x00,

changelog.txt
libxm7 Changelog

----- version release 1.06 ------

[bug] Effect E5x (set note finetune) now works correctly also with MODs
[bug] Effect Dxx (pattern break) works as expected when using hex values

----- version release 1.04 ------

[add] MOD support added for 4 and 8 channel modules subtypes:'M.K.','M!K!','FLT4','OCTA','CD81','FLT8'
[add] MOD support added for 1 to 16 channel modules subtypes:'TDZx','xCHN','xxCH'
[add] On-the-fly sample change feature, for ProTracker MOD modules
[add] It's possible to select different replay modes
[add] It's possible to select different panning modes and panning 'apertures'

[XM7_LoadXM() ARM9 XM loading function:]

[bug] XM7_LoadXM() now will load instrument envelopes (both volume&panning) and settings (fadeout/autovibrato)
      even if the file has been saved with SkaleTracker

----- version release 1.00 ------

[add] Instrument panning envelope is now supported
[add] Instrument 'auto' vibrato is now supported
[add] Amiga frequency table mode (MOD native mode) is now supported

[bug] Effect Hxy (slide global volume) now has memory effect
[bug] Effect EDx (note delay) now retrigs last note and his envelope when specified on a 'empty' line
[bug] Effect E9x (note retrig) now retrigs instrument envelope too
[bug] Effect Lxx (Set volume envelope position) now starts working in the first tick
[bug] Instrument finetune parameter 3 LSB are now ignored, as in FastTrackerII

[XM7_LoadXM() ARM9 XM loading function:]

[bug] XM7_LoadXM() now doesn't fail while loading XMs containing multi-sample instrument with 16bits
      samples in it

----- beta version 0.93 ------

[add] Vibrato effect (4xy) and vibrato control effect (E4x) are now supported.
[add] Tremolo effect (7xy) and tremolo control effect (E7x) are now supported.
[add] Vibrato with volume slide effect (6xy) is now fully supported.
[add] Tremor effect (Txy) is now supported.
[add] Pattern loop effect (E6x) is now supported.
[add] Instrument finetune effect (E5x) is now supported.
[add] Glissando control effect (E3x) is now supported.
[add] Vibrato volume column effect (Vx) is now supported.
[add] Set vibrato speed volume column effect (Sx) is now supported.

[bug] notes specified 'alone' (with no istrument) in the pattern now shouldn't lead unpredictable
      results (it was happening only if you never specified an instrument before in that channel)
      Also, notes 'alone' should now be played correctly
[bug] high pitch notes (octave 7 and above) of instruments with sample finetuning (or pitch bended)
      now shouldn't be "muted"
[bug] Instrument envelopes having points over the 255th tick should now work correctly anyway
[bug] effect EEx (pattern delay) now should produce a correct delay even when it used with an Fxx
      effect (set song speed) on the same line
[bug] effect 9xx (sample offset) now shouldn't give weird side effects when used with a sample
      that has a loop... and 'memory effect', should work, now.

[XM7_LoadXM() / XM7_Unload() ARM9 XM loading/unloading functions:]

[bug] "squeezed" XMs can now be loaded correctly
[bug] conversion from ping pong to forward loop now shouldn't cause instrument detuning
[bug] XMs that fail to load now will be correctly unallocated from memory using XM7_Unload()

----- beta version 0.81 ----

[add] Portamento slide effects (1xx, 2xx, E1x, E2x, X1x, X2x) are now supported.
[add] Portamento to note effect (3xx) is now supported.
[add] Portamento to note with volume slide effect (5xy) is now fully supported.
[add] Jump to position effect (Bxx) and Pattern break effect (Dxx) are now supported.
[add] Key off effect (Kxx) is now supported.
[add] Portamento to note volume column effect (Mx) is now supported.

[bug] istruments specified 'alone' (with no note) in the pattern now shouldn't lead to unpredictable
      results even when used when you never specified a note before in that channel.
      Also, they should now work even after key-offs of instrument with no envelopes.
[bug] effect Rxy (retrig note with volume slide) used at the same time when setting a volume on
      the volume column now should correctly reset the volume before every every retrig.
      'Memory effect', should work correctly now, even when only one of the values (x or y) is 0.
[bug] effect 0xy (Arpeggio) now keeps on working even in the 'extra ticks' added by an effect EEx
      (Pattern Delay) on the same line.
[bug] effect 9xx (sample offset) is now not ignored when used with a sample that has a loop.
[bug] libxm7 startSound() (internal) function has been renamed, so you don't have to rename yours.

----- beta version 0.59 ----

(this was the first release)


include/libxm7.h
// SPDX-License-Identifier: MIT
//
// Copyright (c) 2018 sverx

#ifndef LIBXM7_H__
#define LIBXM7_H__

#include <nds/ndstypes.h>

// Error defines:
#define XM7_ERR_NOT_A_VALID_MODULE				0x01
#define XM7_ERR_UNKNOWN_MODULE_VERSION			0x02
#define XM7_ERR_UNSUPPORTED_NUMBER_OF_CHANNELS	0x03
#define XM7_ERR_UNSUPPORTED_PATTERN_HEADER		0x08
#define XM7_ERR_INCOMPLETE_PATTERN				0x09
#define XM7_ERR_UNSUPPORTED_INSTRUMENT_HEADER	0x10
#define XM7_ERR_NOT_ENOUGH_MEMORY				0x100

// Status bits defines:
#define XM7_STATE_EMPTY							0x0000
#define XM7_STATE_READY                         0x4000
#define XM7_STATE_STOPPED                       XM7_STATE_READY
#define XM7_STATE_PLAYING                       0x6000
#define XM7_STATE_ERROR							0x8000
// (in case of error in loading the module, the 15th bit will be set and
//    lower bits will contain the error number as defined above)

// Panning type defines:
#define XM7_PANNING_TYPE_NORMAL  				0x00
#define XM7_PANNING_TYPE_AMIGA       			0x01

// Panning displacement defines:
#define XM7_HARD_PANNING_DISPLACEMENT           0
#define XM7_MONO_PANNING_DISPLACEMENT           64
#define XM7_DEFAULT_PANNING_DISPLACEMENT        42
// 42 = 127/3 circa = panning 1/3 + 2/3

// Replay style flags:
#define XM7_REPLAY_STYLE_XM_PLAYER              0x00
#define XM7_REPLAY_STYLE_MOD_PLAYER             0x01
#define XM7_REPLAY_ONTHEFLYSAMPLECHANGE_FLAG    0x02

// Replay style default defines:
#define XM7_REPLAY_STYLE_FT2                    0x00
#define XM7_REPLAY_STYLE_PT                     (XM7_REPLAY_STYLE_MOD_PLAYER|XM7_REPLAY_ONTHEFLYSAMPLECHANGE_FLAG)

typedef struct {
	u8 Note;                // 0 = no note; 1..96 = C-0...B-7; 97 = key off
	u8 Instrument;          // 0 or 1..128
	u8 Volume;
	u8 EffectType;
	u8 EffectParam;
}__attribute__ ((packed)) XM7_SingleNote_Type;

typedef struct {
    XM7_SingleNote_Type Noteblock[1];
}__attribute__ ((packed)) XM7_SingleNoteArray_Type;

typedef struct {
  s8 Data[1];
}__attribute__ ((packed)) XM7_SampleData_Type;

typedef struct {
  s16 Data[1];
}__attribute__ ((packed)) XM7_SampleData16_Type;

typedef struct {

   union {
   XM7_SampleData_Type*    SampleData;
   XM7_SampleData16_Type*  SampleData16;
   };
   
   u32 Length;            // length of data, always in bytes
   
   u32 LoopStart;
   u32 LoopLength;
   
   char Name[22];
   
   u8 Volume;
   u8 Panning;            // (0..0xFF, left to right. 0x80 = center )
   
   s8 RelativeNote;      //  (note relative to C-4)
   s8 FineTune;          //  (finetune, in 128th of an half-tone)
   
   u8 Flags;             //  bit 0: it has a loop
                         //  bit 4: it's a 16 bit sample
	
} XM7_Sample_Type;

typedef struct {

  u16 x;
  u16 y;

} XM7_EnvelopePoints_Type;

typedef struct {

    XM7_Sample_Type* Sample[16];      // pointers to the samples
    
	XM7_EnvelopePoints_Type VolumeEnvelopePoint[12];   //  -- Envelope points: x,y...x,y.... 
	XM7_EnvelopePoints_Type PanningEnvelopePoint[12];  //  -- in couples (2 words/point => a maximum of 12 points).
    
    u32 VibratoSweep;              //  0..0x10000
	
	u16 VolumeFadeout;
	
	char Name[22];
	
	u8 SampleforNote[96];             // sample number for note 1..96
	
	u8 NumberofSamples;               // 0..16  (zero... no samples!)
	
	u8 NumberofVolumeEnvelopePoints;
	u8 NumberofPanningEnvelopePoints;
	
	u8 VolumeSustainPoint;
	u8 VolumeLoopStartPoint;
	u8 VolumeLoopEndPoint;
	
	u8 PanningSustainPoint;
	u8 PanningLoopStartPoint;
	u8 PanningLoopEndPoint;
	
	u8 VolumeType;                 //  bit 0: On; 1: Sustain; 2: Loop
	u8 PanningType;                //  bit 0: On; 1: Sustain; 2: Loop
	
	u8 VibratoType;                //  0..3 (wave type)
	u8 VibratoDepth;               //  0..0x0F
    u8 VibratoRate;                //  0..0xFF  ( 0..0x3F on FT2)

} XM7_Instrument_Type;

typedef struct {

    // status of the engine/module/playback
	u16 State;               // bit 15: error if set
                             // bit 14: module loaded if set
                             // bit 13: module playing if set
	u16 ModuleLength;
	u16 RestartPoint;
    u16 NumberofPatterns;     // ...up to 256
    
    // -

	u8 NumberofChannels;      // 1..16 ok, > 16 NO!
	u8 NumberofInstruments;   // 1..128
	u8 FreqTable;             // Frequency table  ( 0= Amiga , 1= linear )
	u8 DefaultTempo;          // 1..31  (ticks per 'line', default 6)
    
    // -
    
	u8 DefaultBPM;          // 32..255 (BeatPerMinute)

	u8 AmigaPanningEmulation;       // should the panning be 'fixed' in the Amiga style?
	u8 AmigaPanningDisplacement;    // amiga panning 'displacement'
	                                // will be added to 0x00 / subtracted to 0x7F

    u8 ReplayStyle;                 // bit 0: 0 = FT2 style
                                    //        1 = MOD player style
                                    // bit 1: on-the-fly sample change flag (1=ON)

    // -
    
   	// these _could_ change during the playback!
	u8 CurrentTempo;        // copy of DefaultTempo (at startup)
	u8 CurrentBPM;          // copy of DefaultBPM   (at startup)
	u8 CurrentGlobalVolume; // 0x40                 (at startup)

	u8 CurrentSongPosition;      //  the pattern in playback now (position in the PatternOrder array)
	u8 CurrentPatternNumber;     //  the pattern in playback now (its number!)
	u8 CurrentLine;              //  the line in playback now (0..length of the pattern)
	u8 CurrentTick;              //  the tick in playback now (0..31)
    
	u8 CurrentDelayLines;        //  the lines to "delay" after this line (effect EEx)
    // -
    
    u16 CurrentDelayTick;        //  the ticks to "delay" after this line (effect EEx)
	u16 CurrentAdditionalTick;   //  the "additional" tick we are in (effect EEx)

    // -
 
	u8 PatternOrder[256];   // play the patterns in this order (PatternNumber is 0..255)

	u16 PatternLength[256];  // the length (in lines) of each pattern (min 1, max 256)  (default=64!)
	
	XM7_SingleNoteArray_Type* Pattern [256];   // pointer to the beginning of every single pattern
	                                           // length = (PatternLength)*(NumberofChannels)*5  (bytes)

    XM7_Instrument_Type* Instrument[128]; // pointer to the instruments
    
    // -

	u8 CurrentSampleVolume[16];     // the volume of the sample on this channel ( 0..0x40 )
	u8 CurrentSamplePanning[16];    // the panning of sample in this channel    ( 0..0xFF left to right, 0x80 = center )
	
	u8 CurrentSampleVolumeEnvelopeState[16];   // the volume envelope state of the sample on this channel
  	u8 CurrentSamplePanningEnvelopeState[16];   // the panning envelope state of the sample on this channel
    
    u8 CurrentSampleVolumeEnvelope[16];        // the envelope volume of the sample on this channel ( 0..0x40 )
    u8 CurrentSamplePanningEnvelope[16];       // the envelope panning of the sample on this channel ( 0..0x40 )

	u16 CurrentSampleVolumeEnvelopePoint[16];  // the volume envelope point (x) of the sample on this channel
	u16 CurrentSamplePanningEnvelopePoint[16]; // the panning envelope point (x) of the sample on this channel

	u16 CurrentSampleVolumeFadeOut[16];        // the fadeout volume of the sample on this channel  ( 0..0x8000 )

	s32 CurrentSamplePortamento[16];     // the period pitch  (for 1xx, 2xx, 3xx, E1x, E2x, X1x, X2x)
	s32 CurrentSamplePortaDest[16];      // the period pitch to achieve (for 3xx and 5xx)
	
	u8 CurrentGlissandoType[16];        // the glissando style (0 = fretless; 1 = round to nearest)
	
	u8 CurrentChannelLastNote[16];       // the number of the last note triggered in this channel        [1 - 96]
	u8 CurrentChannelLastInstrument[16]; // the number of the last instrument triggered in this channel  [1 - 128]
	
	u8 CurrentChannelIsArpeggioedNote[16];  // flag: the note is being arpeggioed
	
	u8 CurrentLoopBegin  [16];               // which line to loop to when meeting E6x effect
	u8 CurrentLoopCounter[16];               // how many times I'm looping
	u8 CurrentLoopEnd    [16];               // the line this looping ends
	
	s8 CurrentVibratoValue [16];            // the vibrato value we should add
	u8 CurrentVibratoType  [16];            // which waveform are we using      [0..2, 4..6]
	u8 CurrentVibratoPoint [16];            // where we are in the vibrato wave [0..63]

	s8 CurrentTremoloVolume[16];          // the volume we should add/sub because of tremolo 
	u8 CurrentTremoloType  [16];          // which tremolo waveform are we using [0..2, 4..6]
	u8 CurrentTremoloPoint [16];          // where we are in the tremolo wave    [0..63]
    
    u8 CurrentTremorPoint  [16];          // where we are in the tremor square wave  [0..x+y+2]
    u8 CurrentTremorMuting [16];          // 1 = Tremor is muting.
    
    u8  CurrentAutoVibratoPoint[16];      // modulator wave is 8 bit for 1 period
    u32 CurrentAutoVibratoSweep[16];      // values 0..0x10000
	
	s8 CurrentFinetuneOverride   [16];    // this is the value for overriding finetune
	u8 CurrentFinetuneOverrideOn [16];    // the flag...
	
	u8 Effect1xxMemory[16];    // the memory for the 1xx effect             [0x00..0xFF]
	u8 Effect2xxMemory[16];    // the memory for the 2xx effect             [0x00..0xFF]
	u8 Effect3xxMemory[16];    // the memory for the 3xx effect (and Mx)    [0x00..0xFF]
	u8 Effect4xxMemory[16];    // the memory for the 4xx effect (and Vx/Sx) [0x00..0xFF]
	u8 Effect7xxMemory[16];    // the memory for the 7xx effect             [0x00..0xFF]
    u8 Effect9xxMemory[16];    // the memory for the 9xx effect (YES! has memory!) [0x00..0xFF]
	u8 EffectAxyMemory[16];    // the memory for the Axy (and 5xx,6xx) effect [0x01..0xF0]
	u8 EffectE1xMemory[16];    // the memory for the E1x effect            [1..0xF]
	u8 EffectE2xMemory[16];    // the memory for the E2x effect            [1..0xF]
	u8 EffectEAxMemory[16];    // the memory for the EAx (and EBx) effect  [1..0xF]
	u8 EffectHxyMemory[16];    // the memory for the Hxy effect            [0x01..0xF0]
	u8 EffectPxyMemory[16];    // the memory for the Pxy effect            [0x01..0xF0]
	u8 EffectRxyMemory[16];    // the memory for the Rxy effect            [0x00..0xFF]
	u8 EffectTxyMemory[16];    // the memory for the Txy effect            [0x11..0xFF]
	u8 EffectX1xMemory[16];    // the memory for the X1x effect            [1..0xF]
	u8 EffectX2xMemory[16];    // the memory for the X2x effect            [1..0xF]
    
    u32 Silence;               // ... a silent 4-bytes sample
    
    // -
    
	char ModuleName  [20];
	char TrackerName [20];

} XM7_ModuleManager_Type;

typedef struct {
	char FixedText[17];    //  ID text: must be 'Extended module: ' ; will be checked
	char XMModuleName[20];
	char FixedChar;        //  MUST be 0x1a                         ; will be checked
	char TrackerName[20];
	u16 Version;           //  MUST be 0x103 or 0x104               ; will be checked
	u32 HeaderSize;
	u16 SongLength;
	u16 RestartPosition;
	u16 NumberofChannels;  //  1..32, > 16 won't be accepted anyway (we've got 16 chn on the DS)
    u16 NumberofPatterns;
	u16 NumberofInstruments;
	u16 XMModuleFlags;
	u16 DefaultTempo;
	u16 DefaultBPM;
	u8 PatternOrder [257]; // 1 more, to build a pointer to this point 
}__attribute__ ((packed)) XM7_XMModuleHeader_Type;

typedef struct {
	u32 HeaderLength;                // _always_ 0x9
	u8 PackingType;                  // MUST be ZERO!
	u16 NumberofLinesinThisPattern;  // 1..256
	u16 PackedPatterndataLength;     // CAN be ZERO!
	u8 PatternData[1];               // I need this to build a pointer to this point
}__attribute__ ((packed)) XM7_XMPatternHeader_Type;

typedef struct {
	u32 InstrumentHeaderLength;
	char Name[22];
	u8 Type;                         // MUST be ZERO! (...but I've seen everything here :| ...)
	u16 NumberofSamples;             // 0..16 ... oh yes, ...it CAN be ZERO...
	u8 NextHeaderPart [1];           // I need this to build a pointer to this point
}__attribute__ ((packed)) XM7_XMInstrument1stHeader_Type;

typedef struct {
	u32 SampleHeaderLength;
	u8  SampleforNotes[96];          // the sample to be used for each note
	u16 VolumeEnvelopePoints [24];
	u16 PanningEnvelopePoints[24];   // Envelope points: x,y...x,y.... in couples 
	                                 // (2 halfwords/point => a maximum of 12 points).
	u8 NumberofVolumePoints;
	u8 NumberofPanningPoints;

	u8 VolumeSustainPoint;
	u8 VolumeLoopStartPoint;
	u8 VolumeLoopEndPoint;

	u8 PanningSustainPoint;
	u8 PanningLoopStartPoint;
	u8 PanningLoopEndPoint;
	
	u8 VolumeType;                   // bit 0: On; 1: Sustain; 2: Loop
	u8 PanningType;                  // bit 0: On; 1: Sustain; 2: Loop
	
	u8 VibratoType;
	u8 VibratoSweep;
	u8 VibratoDepth;
	u8 VibratoRate;
	
	u16 VolumeFadeOut;
	u16 Reserved[11];               // no use

    u8 NextDataPart[1];            // I need this to build a pointer to this point
}__attribute__ ((packed)) XM7_XMInstrument2ndHeader_Type;

typedef struct {
	u32 Length;
	u32 LoopStart;
	u32 LoopLength;
	
	u8 Volume;
	s8 FineTune;
	u8 Type;                        // bit 0,1 - 0: no loop; 1: forward; 2:ping-pong
	                                // bit   4 - 0: 8 bit; 1: 16 bit
	u8 Panning;						// 0x00..0xff
	s8 RelativeNote;				// -128..+127
	u8 Reserved;					// no use

	char Name[22];
	
	u8 NextHeader[1];              // I need this to build a pointer to this point
}__attribute__ ((packed)) XM7_XMSampleHeader_Type;

// XM file format was found at http://aluigi.altervista.org/mymusic/xm.txt


// MOD section follows:

typedef struct {
	char Name[22];           //  Sample name
	u16 Length;              //  Sample lenght in Motorola words (*)
	u8 FineTune;             //  Finetune, signed nibble.
	u8 Volume;               //  0..0x40
    u16 LoopStart;           //  Sample repeat offset in Motorola words (*)
    u16 LoopLength;          //  Sample repeat length in Motorola words (*)
                             //  (sample has loop if this is > 2 )
}__attribute__ ((packed)) XM7_MODSampleHeader_Type;

typedef struct {
	char MODModuleName[20];                   // MODule name
    XM7_MODSampleHeader_Type Instrument[31];  // array of 31 instruments
    u8 SongLength;                            // MODule lenght in POT
    u8 RestartPosition;                       // MODule restart position in POT
    u8 PatternOrder[128];                     // POT
    char FileFormat[4];                       //    --> SEE NOTE!
	u8 NextDataPart;                          // need this to build a pointer to this point 
}__attribute__ ((packed)) XM7_MODModuleHeader_Type;

typedef struct {
  u8 PeriodH;          // high nibble of period (low nibble) & instrument MSB (high nibble)
  u8 PeriodL;          // low byte of period
  u8 Instr_EffType;    // instrument (high nibble) & effect type (low nibble)
  u8 EffParam;         // effect param
}__attribute__ ((packed)) XM7_MODSingleNote_Type;

typedef struct {
  XM7_MODSingleNote_Type SingleNote[1];      // 
}__attribute__ ((packed)) XM7_MODPattern_Type;

// (*) Motorola words = 16 bit BIG ENDIAN = 1 Motorola word is 2 bytes, byte-swapped

// end of MOD section


// ARM7 functions
void XM7_Initialize        (void);
void XM7_PlayModule        (XM7_ModuleManager_Type* TheModule);
void XM7_PlayModuleFromPos (XM7_ModuleManager_Type* TheModule, u8 position);
void XM7_StopModule        (void);


// ARM9 functions (... well, you can use them even on ARM7 if you want...)
u16  XM7_LoadXM    (XM7_ModuleManager_Type*, XM7_XMModuleHeader_Type*);
void XM7_UnloadXM  (XM7_ModuleManager_Type*);

u16  XM7_LoadMOD   (XM7_ModuleManager_Type*, XM7_MODModuleHeader_Type*);
void XM7_UnloadMOD (XM7_ModuleManager_Type*);

void XM7_SetReplayStyle  (XM7_ModuleManager_Type* Module, u8 style);
void XM7_SetPanningStyle (XM7_ModuleManager_Type* Module, u8 style, u8 displacement);

#endif // LIBXM7_H__
